File: br-ingest-linear/src/linear.ts
import fetch from 'node-fetch';

export type LinearIssue = {
  id: string;
  number: number;
  identifier: string;
  title: string;
  description?: string | null;
  createdAt: string;
  updatedAt: string;
  completedAt?: string | null;
  priority?: number | null;
  estimate?: number | null;
  assignee?: { name?: string | null } | null;
  state?: { name?: string | null; type?: string | null } | null;
  team?: { key?: string | null; name?: string | null } | null;
  project?: { name?: string | null } | null;
  labels?: { nodes?: Array<{ name?: string | null }> } | null;
};

type PageInfo = { hasNextPage: boolean; endCursor?: string | null };

export const ISSUES_Q = `
query Issues($teamKey:String!, $updatedAfter:DateTime, $after:String) {
  issues(
    filter:{ team:{ key:{ eq:$teamKey } }, updatedAt:{ gt:$updatedAfter } },
    first: 50, after: $after
  ) {
    nodes {
      id number identifier title description
      createdAt updatedAt completedAt
      priority estimate
      assignee { name }
      state { name type }
      team { key name }
      project { name }
      labels { nodes { name } }
    }
    pageInfo { hasNextPage endCursor }
  }
}`;

export async function* iterIssues(token: string, teamKey: string, updatedAfterISO: string) {
  let after: string | null = null;
  while (true) {
    const res = await fetch('https://api.linear.app/graphql', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ query: ISSUES_Q, variables: { teamKey, updatedAfter: updatedAfterISO, after } }),
    });
    if (!res.ok) throw new Error(`Linear ${res.status}`);
    const body: any = await res.json();
    const nodes: LinearIssue[] = body?.data?.issues?.nodes ?? [];
    const page: PageInfo = body?.data?.issues?.pageInfo ?? { hasNextPage: false };
    yield nodes;
    if (!page.hasNextPage) break;
    after = page.endCursor ?? null;
  }
}

File: sync_connectors/connectors/linear.py
"""Linear connector implementation."""
from __future__ import annotations

from typing import Optional

from .base import ConnectorBase, ConnectorError


class LinearConnector(ConnectorBase):
    name = "Linear"
    token_env_var = "LINEAR_API_KEY"

    def _sync_impl(self, token: str) -> tuple[str, Optional[int]]:
        headers = {
            "Authorization": token,
            "Content-Type": "application/json",
        }
        query = (
            "query LinearTeams($after: String) {"
            " viewer { name displayName }"
            " teams(first: 50, after: $after) {"
            "   nodes { id name archivedAt }"
            "   pageInfo { hasNextPage endCursor }"
            " }"
            "}"
        )
        active_count = 0
        account_name: Optional[str] = None
        cursor: Optional[str] = None
        while True:
            payload = {"query": query, "variables": {"after": cursor}}
            response = self._request_with_retries(
                "POST",
                "https://api.linear.app/graphql",
                headers=headers,
                json=payload,
            )
            if "errors" in response:
                raise ConnectorError(str(response["errors"]))
            viewer = response.get("data", {}).get("viewer", {})
            account_name = (
                viewer.get("name")
                or viewer.get("displayName")
                or account_name
                or "Linear User"
            )
            teams = response.get("data", {}).get("teams", {})
            nodes = teams.get("nodes", [])
            active_count += sum(1 for node in nodes if not node.get("archivedAt"))
            page_info = teams.get("pageInfo", {})
            if page_info.get("hasNextPage"):
                cursor = page_info.get("endCursor")
            else:
                break
        if account_name is None:
            account_name = "Linear User"
        return account_name, active_count


__all__ = ["LinearConnector"]
